### PR TITLE
feat(float): add fclose command

### LINKED ISSUE #9663: floating windows: Ex commands (:fclose, :fedit, ... or :float)
#6619 added API support for floating windows.

Next step is to add some "chrome" for this.

- ✅ [#25561](https://github.com/neovim/neovim/pull/25561) Some have requested an easier way to close the top (or all) floating windows, without navigating-to the window to `:close` it. We could  introduce the "f" prefix similar to "p" family of commands (`:pclose`, `:ptag`).  
  - `:fclose` closes the highest z-index, and `:[count]fclose` closes window at  `count` z-index.  
  - `:fclose!` closes all floats.
- ~~Or a `:float` modifier? Analogous to `:vertical`, `:tab`, etc.~~ [#12050](https://github.com/neovim/neovim/pull/12050)

Instead of introducing analogs for `:pedit`, `:ptag`, etc., there could be a global option that configures the "preview window" to be a floating window (with  default position, dimensions, etc.).

### PR DESCRIPTION
**Pull request by glepnir**
`:[count]fclose[!]` close `count` floating window by zindex sort or `!` close all floating windows

Fix #9663

### CODE REVIEW

**Comment by justinmk** [comment]
Cool to see this!

- The general idea of #9663 was to use `:pclose` and related preview-window features as inspiration. 
    - Presumably we _don't_ need analogs to `:ptag`/`:psearch`/`:pedit`, instead we should implement Vim's `:help preview-popup` behavior.
- Based on `popup.txt` it looks like Vim doesn't have anything similar to `:fclose` for "popups" so we're on our own here.

**Comment by glepnir** [comment]
yes `zindex` sorting needs fix. I will continue it after the other PR is completed.

File: src/nvim/window.c [hunk 1]
```diff
@@ -7658,3 +7658,29 @@ win_T *lastwin_nofloating(void)
   }
   return res;
 }
+
+static int floating_zindex_compare(const void *a, const void *b)
+{
+  return (*(win_T **)a)->w_float_config.zindex < (*(win_T **)b)->w_float_config.zindex;
```
Comment (bfredl) [inline src/nvim/window.c:7664]: better to be `(*(win_T **)b)->w_float_config.zindex - (*(win_T **)a)->w_float_config.zindex` (return value should be either negative, zero, or positive for all three possible comparisons)

File: src/nvim/ex_docmd.c [hunk 2]
```diff
@@ -7357,6 +7357,13 @@ static void ex_terminal(exarg_T *eap)
   do_cmdline_cmd(ex_cmd);
 }
 
+/// ":fclose"
+static void ex_floatclose(exarg_T *eap)
+{
+  bool bang = eap->cmd[strlen(eap->cmd) - 1] == '!' ? true : false;
+  win_float_remove(bang, eap->line1);
```
Comment (bfredl) [inline src/nvim/ex_docmd.c:7364]: for `bang` can use `eap->forceit`

Comment (glepnir) [reply src/nvim/ex_docmd.c]: oh get it. thanks for tip

File: src/nvim/ex_cmds.lua [hunk 3]
```diff
@@ -1814,6 +1814,12 @@ module.cmds = {
     addr_type='ADDR_NONE',
     func='ex_nohlsearch',
   },
+  {
+    command='fclose',
+    flags=bit.bor(BANG, RANGE),
+    addr_type='ADDR_OTHER',
+    func='ex_floatclose',
+  },
```
Comment (zeertzjq) [inline src/nvim/ex_cmds.lua:1822]: This is a strange place to add this command. It should be next to another command that start with `f`.

Comment (glepnir) [reply src/nvim/ex_cmds.lua]: thanks. Commands are sorted alphabetically. I did not notice T.T

File: runtime/doc/windows.txt [hunk 4]
```diff
@@ -940,6 +940,11 @@ CTRL-W CTRL-Z					*CTRL-W_CTRL-Z* *:pc* *:pclose*
 		window editing it).  The command fails if any "Preview" buffer
 		cannot be closed.  See also |:close|.
 
+							*:fc* *:fclose*
+:[count]fc[lose][!]
+		Close [count]th floating window by zindex order. '!' to closes
```
Comment (zeertzjq) [inline runtime/doc/windows.txt:945]:
```suggestion
		Close [count]th floating window by zindex order. '!' to close
```

File: runtime/doc/index.txt [hunk 5]
```diff
@@ -1455,6 +1455,7 @@ tag		command		action ~
 |:packadd|	:pa[ckadd]	add a plugin from 'packpath'
 |:packloadall|	:packl[oadall]	load all packages under 'packpath'
 |:pclose|	:pc[lose]	close preview window
+|:fclose|	:fc[lose]	close floating window
```
[the remark this answers is not in the document]
Comment (glepnir) [reply runtime/doc/index.txt]: also need between `exusage` and `file` in `ex_cmds.lua` ?

File: scripts/gen_help_html.lua [hunk 6]
```diff
@@ -61,6 +61,7 @@ local new_layout = {
 -- TODO: These known invalid |links| require an update to the relevant docs.
 local exclude_invalid = {
   ["'previewpopup'"] = "quickref.txt",
+  ["[count]fclose[!]"] = "windows.txt",
```
Comment (zeertzjq) [inline scripts/gen_help_html.lua:64]: Remove this. This string doesn't even exist in the docs.

Comment (glepnir) [reply scripts/gen_help_html.lua:64]: if remove this, `help_spec.lua` will through an error.

File: runtime/doc/windows.txt [hunk 7]
```diff
@@ -940,6 +940,11 @@ CTRL-W CTRL-Z					*CTRL-W_CTRL-Z* *:pc* *:pclose*
 		window editing it).  The command fails if any "Preview" buffer
 		cannot be closed.  See also |:close|.
 
+							*:fc* *:fclose*
+:[count]fc[lose][!]
+		Close [count]th floating window by zindex order. '!' to close
+		all floating windows.
+
```
Comment (zeertzjq) [inline runtime/doc/windows.txt:947]: I don't think this blocking is in the correct section. This section is about the preview window.

Comment (glepnir) [reply runtime/doc/windows.txt:947]: we don't have a popup chapter :thinking:

File: scripts/gen_help_html.lua (hunk 6 above)
Comment (zeertzjq) [reply scripts/gen_help_html.lua:64]: I don't think that's possible, and I don't get an error.

Comment (glepnir) [reply scripts/gen_help_html.lua:64]: you are right . now it works fine. before always tell me invalid tags

File: runtime/doc/windows.txt (hunk 7 above)
Comment (glepnir) [reply runtime/doc/windows.txt]: okay

**Comment by bfredl** [comment]
Possible follow up: We could maintain a vector over zindex sorted floats in `window.c`. similar to `layers` in `ui_compositor.c` but only including floats and always available (also when ext_multigrid is used and compositor disabled) . Then we won't need the `qsort()` in the command  implementation (and also this info could be used elsewhere, i e be queried by some API function).

File: src/nvim/ex_docmd.c [hunk 8]
```diff
@@ -7357,6 +7357,12 @@ static void ex_terminal(exarg_T *eap)
   do_cmdline_cmd(ex_cmd);
 }
 
+/// ":fclose"
+static void ex_floatclose(exarg_T *eap)
```
Comment (justinmk) [inline src/nvim/ex_docmd.c:7361]: Seems more conventional to name the `ex_` function using the same name as the command: `ex_fclose`

File: test/functional/ui/float_spec.lua [hunk 9]
```diff
@@ -10927,6 +10927,186 @@ describe('float window', function()
...
+          {5:│}{2:~       }{5:│}|
+          {5:└────────┘}|
+        ## grid 6
+          {5:┌────┐}|
+          {5:│}{1:    }{5:│}|
+          {5:│}{2:~   }{5:│}|
+          {5:│}{2:~   }{5:│}|
+          {5:│}{2:~   }{5:│}|
+          {5:└────┘}|
+        ]],float_pos= expected_pos}
```
Comment (justinmk) [inline test/functional/ui/float_spec.lua:11003]: Is the `grid` arg doing anything important here or could we save a lot of noise in the test by only specifying `float_pos=expected_pos`

File: test/functional/ui/float_spec.lua [hunk 10]
```diff
@@ -10927,6 +10927,186 @@ describe('float window', function()
         ]])
       end
     end)
+
+    it('fclose command #9663', function()
+      local bufA = meths.create_buf(false,false)
+      local bufB = meths.create_buf(false,false)
+      local bufC = meths.create_buf(false,false)
+      local bufD = meths.create_buf(false,false)
+      local config_A = {relative='editor', width=11, height=11, row=5, col=5, border ='single', zindex=50}
```
Comment (justinmk) [inline test/functional/ui/float_spec.lua:10936]: doesn't follow our naming conventions

```
buf_d
config_a
```

File: runtime/doc/windows.txt [hunk 11]
```diff
@@ -392,6 +392,11 @@ CTRL-W CTRL-O					*CTRL-W_CTRL-O* *:on* *:only*
 		given, then they become hidden.  But modified buffers are
 		never abandoned, so changes cannot get lost.
 
+							*:fc* *:fclose*
+:[count]fc[lose][!]
+		Close [count]th floating window by zindex order. '!' to close
+		all floating windows.
```
Comment (justinmk) [inline runtime/doc/windows.txt:398]: "order" seems to be the key word. It's not obvious that the zindex value itself is only relevant _relative_ to other zindexes. So if there are zindexes 99, 999, 9999, 

- `:2fclose` closes two windows (not obvious _which_ windows).
- `:fclose` closes all floats if cursor is _not_ in a float
- `:fclose` closes the highest-zindex float if cursor is in a float

File: src/nvim/ex_docmd.c (hunk 8 above)
Comment (glepnir) [reply src/nvim/ex_docmd.c:7361]: before think about this. because C has `fclose`

File: test/functional/ui/float_spec.lua (hunk 9 above)
Comment (glepnir) [reply test/functional/ui/float_spec.lua:11003]: before has `expected_pos[6] = nil` it's vaidate float window pos after close

**Pull request merged**